### ISSUE TITLE
StreamDefinitionToDslConverter may create unparsable dsl from valid definition

### ISSUE THREAD

**Issue by bah91** [spring-cloud/spring-cloud-dataflow#2033]
I have encountered an Exception on deploying a stream containing a jdbc source.

The stream definition looks similar to this:
 
jdbc-mssql --query='UPDATE top (100) ASSURANCE SET assurance_flag = 1 OUTPUT Inserted.* WHERE assurance_flag IS NULL' --password='******' --username='******' --url='jdbc:sqlserver://db:1433;databaseName=Spring' --cron='*/10 * * * * *' --max-messages=-1 | cust-processor | router --default-output-channel=out

On deployment the following exception is thrown: 

org.springframework.cloud.dataflow.core.dsl.CheckPointedParseException: 100E:(pos x): Found unexpected data after stream definition: ';'

It appears it does not like the ';' symbol from the jdbc URL. 

I am using Spring Cloud Dataflow v1.3.0.RELEASE with Skipper v1.0.0.RELEASE. The same definitions worked fine in v1.2.0.RELEASE

**Comment by jorgerod** [comment]
+1 
The problem is with character ';'

I have the stream:
`stream create --name stream-1 --definition 'http-source-kafka --server.port=9900 | couchbase-sink-kafka --inputType="application/x-java-object;type=com.example.dto.InputDto"'`

This work but on deployment:
`org.springframework.cloud.dataflow.core.dsl.CheckPointedParseException: 100E:(pos 92): Found unexpected data after stream definition: ';'`

The problemn is with ";" in this part: application/x-java-object;type=com.example.dto.InputDto

I am using Spring Cloud Dataflow v1.3.0.RELEASE with Skipper v1.0.0.RELEASE. The same definitions worked fine in v1.2.0.RELEASE

**Comment by sabbyanandan** [comment]
Hi, @bah91 / @jorgerod. Sorry for the delay. Could one of you share the stacktrace of the full exception? Also, what version of the Spring Cloud Stream are you using? It'd be good review a reproducible sample of it.

**Comment by olegz** [comment]
@bah91 @jorgerod I have seen similar parsing issues. Could you please try to supply this property via Properties file? Just wanted to confirm what I believe is a clip parsing issue

**Comment by sabbyanandan** [comment]
I tried to reproduce this problem on 1.2.3 and 1.3.1. With similar stream definitions, I'm unable to see the specific exception. Also, as an FYI, the `StreamParser` (through `UNEXPECTED_DATA_AFTER_STREAMDEF ` condition) remains unmodified functionally since March 2017.  Perhaps there's 

@bah91 @jorgerod: Given are both seeing the same error; we are definitely interested in reviewing more logs; specifically, where do you see the error? In Server or Shell and at what event?

**Comment by jvalkeal** [comment]
I was able to reproduce this and happens with all recent versions. It's related to skipper as doesn't fail if skipper mode is not enabled.

**Comment by jvalkeal** [comment]
Exact stack is:
```
2018-03-07 12:40:09.212 ERROR 18082 --- [nio-9393-exec-7] o.s.c.d.s.c.RestControllerAdvice         : Caught exception while handling a request

org.springframework.cloud.dataflow.core.dsl.CheckPointedParseException: 100E:(pos 85): Found unexpected data after stream definition: ';'
time | log --spring.cloud.stream.bindings.input.contentType=application/x-java-object;type=com.example.dto.InputDto
                                                                                     *
	at org.springframework.cloud.dataflow.core.dsl.Tokens.raiseException(Tokens.java:276) ~[spring-cloud-dataflow-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.core.dsl.StreamParser.eatStream(StreamParser.java:167) ~[spring-cloud-dataflow-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.core.dsl.StreamParser.parse(StreamParser.java:74) ~[spring-cloud-dataflow-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.core.StreamDefinition.<init>(StreamDefinition.java:71) ~[spring-cloud-dataflow-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.server.service.impl.DefaultSkipperStreamService.updateStreamDefinitionFromReleaseManifest(DefaultSkipperStreamService.java:157) ~[spring-cloud-dataflow-server-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.server.service.impl.DefaultSkipperStreamService.doDeployStream(DefaultSkipperStreamService.java:115) ~[spring-cloud-dataflow-server-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.server.service.impl.AbstractStreamService.deployStream(AbstractStreamService.java:80) ~[spring-cloud-dataflow-server-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.dataflow.server.controller.StreamDeploymentController.deploy(StreamDeploymentController.java:150) ~[spring-cloud-dataflow-server-core-1.4.0.BUILD-SNAPSHOT.jar!/:1.4.0.BUILD-SNAPSHOT]
```

**Comment by olegz** [comment]
This is what I've been saying all along. We have some issue with how we parse properties. The above `application/x-java-object;type=com.example.dto.InputDto` is completely valid and in fact would raise issue in SCSt if the `type=` part is missing since it is required. But in any event there is a problem in hop we parse properties as we assume simple `key-value` which is not true for many properties.

**Comment by jvalkeal** [comment]
It gets wrong in 
```
String dslText = new StreamDefinitionToDslConverter().toDsl(updatedStreamAppDefinitions);
```
If I do something like
```
stream create --name foostream --definition 'time | log --inputType="application/x-java-object;type=com.example.dto.InputDto"'
```
What gets passed in into controller is
```
time | log --inputType="application/x-java-object;type=com.example.dto.InputDto"
```
Then `StreamDefinitionToDslConverter` mangles it and creates
```
time | log --spring.cloud.stream.bindings.input.contentType=application/x-java-object;type=com.example.dto.InputDto
```
We miss quoting so parser breaks.

**Comment by jvalkeal** [comment]
This is the converter @tzolov made and he had quite a bit of issues with it.

**Comment by sabbyanandan** [comment]
@jvalkeal: I'm glad you were able to reproduce it. I should have also tried w/ Skipper and not Classic alone. This converter is in place to reverse engineer the stream definition and the associated app/deployment properties based on the stream-manifest persisted in Skipper. 

We rely on the converter a lot more in 1.3 and onwards for several downstream improvements (stream-info, validations, app un-registration etc.,). We will have to make sure to cover all those scenarios after the change.

@tzolov: What are your thoughts?

**Comment by tzolov** [comment]
We faced already a similar issue: #1984. There we introduced property auto-quoting for properties contains certain characters such as spaces. Quick fix is to add `;` to characters list. 

Unfortunately with current App definition API, the information if the original property definition was quoted or not is not preserved. So during the reverse-engineering we don't know if the original property used quotes or not. 

Possible long-term solutions are (1) to extend the App Definition API to keep the property quoting meta-info and use it for reverse engineering or (2) as @jvalkeal has suggested to auto-quote all properties that don't have quotes in their content. 
@sabbyanandan I will resolve the imminent problem of handling the `;` character and we can open another issue for the long-term solution?
